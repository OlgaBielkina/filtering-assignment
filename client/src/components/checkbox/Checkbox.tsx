import React, { useCallback } from 'react';
import classnames from 'classnames';
import styles from './Checkbox.module.scss';

export interface Props {
    label: string;
    checked?: boolean;
    onChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
    className?: string;
}

export const Checkbox: React.FC<Props> = ({ label, checked = false, onChange, className }) => {
    const handleClick = useCallback((event: React.MouseEvent<HTMLLabelElement>) => {
        event.stopPropagation();
    }, []);

    return (
        <label className={classnames(styles.root, className)} onClick={handleClick}>
            <input type="checkbox" checked={checked} onChange={onChange} className={styles.input} />
            <span className="label">{label}</span>
        </label>
    );
};
