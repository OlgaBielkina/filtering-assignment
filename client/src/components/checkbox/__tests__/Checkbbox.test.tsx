import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { Checkbox, Props } from '../Checkbox';

describe('Checkbox', () => {
    const props: Props = {
        label: 'test label',
        checked: true,
        onChange: jest.fn(),
        className: 'testClass',
    };

    it('should render correctly', () => {
        const { asFragment } = render(<Checkbox {...props} />);

        expect(asFragment()).toMatchSnapshot();
    });

    it('should check chekbox by pressing on label', () => {
        const { getByLabelText, getByRole } = render(<Checkbox {...props} />);

        fireEvent.click(getByLabelText(props.label));

        getByRole('checkbox', { checked: true });
    });

    it('should call an onChnage callback when checkbox is checked', () => {
        const { getByLabelText } = render(<Checkbox {...props} />);

        fireEvent.click(getByLabelText(props.label));

        expect(props.onChange).toBeCalledTimes(1);
    });

    it('should prevent click event propagation', () => {
        const onButtonClick = jest.fn();
        const { getByLabelText } = render(
            <button onClick={onButtonClick}>
                <Checkbox {...props} />
            </button>,
        );

        fireEvent.click(getByLabelText(props.label));

        expect(props.onChange).toBeCalledTimes(1);
        expect(onButtonClick).not.toBeCalled();
    });
});
