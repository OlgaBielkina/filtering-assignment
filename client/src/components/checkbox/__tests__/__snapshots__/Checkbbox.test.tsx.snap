// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox should render correctly 1`] = `
<DocumentFragment>
  <label
    class="root testClass"
  >
    <input
      checked=""
      class="input"
      type="checkbox"
    />
    <span
      class="label"
    >
      test label
    </span>
  </label>
</DocumentFragment>
`;
